cli/cmd/gateway/read/command.go

```diff
@@ -0,0 +1,299 @@
+package read
+
+import (
+	"archive/zip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+
+	helmcli "helm.sh/helm/v3/pkg/cli"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gwv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+	gwv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+	"sigs.k8s.io/yaml"
+
+	"github.com/hashicorp/consul-k8s/cli/common"
+	"github.com/hashicorp/consul-k8s/cli/common/flag"
+	"github.com/hashicorp/consul-k8s/cli/common/terminal"
+)
+
+type Command struct {
+	*common.BaseCommand
+
+	kubernetes client.Client
+	restConfig *rest.Config
+
+	set *flag.Sets
+
+	flagGatewayKind      string
+	flagGatewayNamespace string
+	flagKubeConfig       string
+	flagKubeContext      string
+	flagOutput           string
+
+	gatewayName string
+
+	initOnce sync.Once
+	help     string
+}
+
+func (c *Command) Help() string {
+	c.initOnce.Do(c.init)
+	return fmt.Sprintf("%s\n\nUsage: consul-k8s gateway read <gateway-name> [flags]\n\n%s", c.Synopsis(), c.help)
+}
+
+func (c *Command) Synopsis() string {
+	return "Inspect the configuration for a given Gateway."
+}
+
+// init establishes the flags for Command
+func (c *Command) init() {
+	c.set = flag.NewSets()
+
+	f := c.set.NewSet("Command Options")
+	f.StringVar(&flag.StringVar{
+		Name:    "namespace",
+		Target:  &c.flagGatewayNamespace,
+		Usage:   "The namespace of the Gateway to read",
+		Aliases: []string{"n"},
+	})
+	f.StringVar(&flag.StringVar{
+		Name:    "output",
+		Target:  &c.flagOutput,
+		Usage:   "Output the Gateway configuration as 'json' in the terminal or 'archive' as a zip archive in the current directory named after the Gateway.",
+		Default: "archive",
+		Aliases: []string{"o"},
+	})
+
+	f = c.set.NewSet("Global Options")
+	f.StringVar(&flag.StringVar{
+		Name:    "kubeconfig",
+		Aliases: []string{"c"},
+		Target:  &c.flagKubeConfig,
+		Usage:   "Set the path to kubeconfig file.",
+	})
+	f.StringVar(&flag.StringVar{
+		Name:   "context",
+		Target: &c.flagKubeContext,
+		Usage:  "Set the Kubernetes context to use.",
+	})
+
+	c.help = c.set.Help()
+}
+
+// Run runs the command
+func (c *Command) Run(args []string) int {
+	c.initOnce.Do(c.init)
+	c.Log.ResetNamed("read")
+	defer common.CloseWithError(c.BaseCommand)
+
+	if len(args) < 1 || strings.HasPrefix(args[0], "-") {
+		c.UI.Output("Usage: gateway read <gateway-name>")
+		return 1
+	}
+
+	if len(args) > 1 {
+		if err := c.set.Parse(args[1:]); err != nil {
+			c.UI.Output(err.Error(), terminal.WithErrorStyle())
+			return 1
+		}
+	}
+
+	c.gatewayName = args[0]
+
+	if err := c.initKubernetes(); err != nil {
+		c.UI.Output(err.Error(), terminal.WithErrorStyle())
+		return 1
+	}
+
+	if err := c.fetchCRDs(); err != nil {
+		c.UI.Output(err.Error(), terminal.WithErrorStyle())
+		return 1
+	}
+
+	return 0
+}
+
+func (c *Command) fetchCRDs() error {
+	// Fetch Gateway
+	var gateway gwv1beta1.Gateway
```

Comment:
is it possible that someone is running the v1alpha1 of the gateway if they're on an old consul-k8s version? should we worry about that/care about that?

Reply:
v1.2 which introduced `Gateway` to consul-k8s itself consumes `v1beta1.Gateway` [here](https://github.com/hashicorp/consul-k8s/blob/v1.2.9/control-plane/api-gateway/controllers/gateway_controller.go#L402), so customers would have to be using the old standalone consul-api-gateway project in order to be running the `v1alpha1.Gateway`.

Based on that, I don't think we need to support it here.